src/GraphCoarsenBuilder.cpp

```diff
@@ -201,6 +260,239 @@ std::unique_ptr<mfem::SparseMatrix> AssembleMBuilder::GetCoarseM()
+                mfem::DenseMatrix P_EFprime(local_fine_dofs.Size(), local_coarse_dofs_prime.Size());
+                Pedges_.GetSubMatrix(local_fine_dofs, local_coarse_dofs_prime, P_EFprime);
+                comp_EF_EF_.push_back(RTP(P_EF, P_EFprime));
+                auto pair = std::make_pair(face, faceprime);
+                face_pair_to_index[pair] = counter++;
+            }
+        }
+    }
+
+    // EF_E block and E_E block
+    comp_E_E_.resize(num_aggs);
+    for (int agg = 0; agg < num_aggs; ++agg)
+    {
+        GetCoarseAggDofs(agg, local_coarse_dofs);
+        if (local_coarse_dofs.Size() == 0)
+        {
+            mfem::DenseMatrix empty(0, 0);
+            comp_E_E_[agg] = empty;
+            for (int af = 0; af < local_faces.Size(); ++af)
+            {
+                comp_EF_E_.push_back(empty);
+            }
+        }
+        else
+        {
+            GetTableRowCopy(topology_.Agg_edge_, agg, local_fine_dofs);
+            mfem::DenseMatrix P_E(local_fine_dofs.Size(), local_coarse_dofs.Size());
+            Pedges_.GetSubMatrix(local_fine_dofs, local_coarse_dofs, P_E);
+            comp_E_E_[agg] = RTP(P_E, P_E);
+            GetTableRowCopy(topology_.Agg_face_, agg, local_faces);
+            for (int af = 0; af < local_faces.Size(); ++af)
+            {
+                int face = local_faces[af];
+                GetCoarseFaceDofs(face, local_coarse_dofs);
+                mfem::DenseMatrix P_EF(local_fine_dofs.Size(), local_coarse_dofs.Size());
+                Pedges_.GetSubMatrix(local_fine_dofs, local_coarse_dofs, P_EF);
+                // comp_EF_E[index] = RTP(P_EF, P_E);
+                comp_EF_E_.push_back(RTP(P_EF, P_E));
+                // also store transpose, or just have it implicitly?
+            }
+        }
+    }
+}
+
+/// this shares a lot of code with BuildComponents, but I'm not sure it makes
+/// sense to combine them in any way.
+std::unique_ptr<mfem::SparseMatrix> CoefficientMBuilder::GetCoarseM()
+{
+    // what we're doing:
+    //   1. assemble from components, identity weights, compare to original
+    //   2. change interface so we can actually change weights
+    //   3. what to do in parallel....
+
+    // ---
+    // build the components...
+    // ---
+    BuildComponents();
+
+    // temporarily just use hard-coded weights = 1
+    const int num_aggs = topology_.Agg_face_.Height();
+    const int num_faces = topology_.Agg_face_.Width();
+    mfem::Vector agg_weights(num_aggs);
+    agg_weights = 1.0;
+    SetCoefficient(agg_weights);
+
+    // ---
+    // assemble from components...
+    // ---
+    auto CoarseM = make_unique<mfem::SparseMatrix>(
+                       total_num_traces_ + ncoarse_vertexdofs_ - num_aggs,
+                       total_num_traces_ + ncoarse_vertexdofs_ - num_aggs);
+
+    // F_F block, the P_F^T M_F P_F pieces (this explicit transpose is not my favorite)
+    auto face_Agg = smoothg::Transpose(topology_.Agg_face_);
```

Comment:
`GraphTopology` has `face_Agg_` ? Unless that's some how different.

Reply:
You're right, not sure how I missed that. Fixed in last commit.